src/app/epics/user-login.epic.spec.ts

```diff
@@ -0,0 +1,176 @@
+import 'rxjs/add/observable/throw';
+
+import 'rxjs/add/observable/empty';
+import 'rxjs/add/operator/switchMap';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/catch';
+
+import { AnyAction } from 'redux';
+import { IUser } from '../models/user.model';
+import { UserActions } from '../actions/user.actions';
+
+let mockUser: IUser = { name: 'MOCK_user' };
+
+class MockLoginService {
+    login = (username, pass) => {};
+
+    logout = () => {};
+};
+// TODO: Don't feel confident in the testing of Observables.
+// TODO: Needs to be updated since Login Service has changed its error handle.
+//      And because there is now a Router dependency
+class MockUserActions {
+    loginSuccess = (user, date) => ({ type: 'MOCK_LOGIN_SUCCESS', user, date });
+    loginFailed = (err) => ({ type: 'MOCK_LOGIN_ERROR', err });
+};
+
+class MockRouter {
+    router = {
+        navigate: (commands, extras) => {},
+    };
+}
+
+
+describe('UserLoginEpic', () => {
+    let mockUserActions: MockUserActions;
+    let mockLoginService: MockLoginService;
+    let mockRouter: MockRouter;
+    let userLoginEpic: UserLoginEpic;
+    let action$: any;
+
+    
+
+    describe('login', () => {
+        beforeEach(() => {
+            mockLoginService = new MockLoginService();
+            mockUserActions = new MockUserActions();
+            mockRouter = new MockRouter();
+            userLoginEpic = new UserLoginEpic(
+                mockLoginService as any, 
+                mockUserActions as any,
+                mockRouter as any
+            );
+        });
+
+        it('should initiate login and succeed', (done: DoneFn) => {
+            const user: string = 'USER';
+            const pass: string = 'PASSWORD';
+            const date: number = Date.now();
+
+            spyOn(mockLoginService, 'login')
+                .and
+                .returnValue( Observable.of(mockUser) );
+            spyOn(mockUserActions, 'loginSuccess')
+                .and
+                .returnValue( mockUserActions.loginSuccess(user, date) );
+            spyOn(mockUserActions, 'loginFailed')
+                .and
+                .returnValue( mockUserActions.loginFailed('Mock Error') )
+
+            action$ = ActionsObservable.of({ 
+                type: UserActions.LOGIN_USER_REQUEST,
+                username: user,
+                pass: pass,
+            });
+
+            userLoginEpic.login(action$)
+                .subscribe(
+                    () => {},
+
+                    () => done.fail('User Login Epic: login fail'),
+
+                    () => { 
+                        expect( mockLoginService.login )
+                            .toHaveBeenCalledWith(user, pass);
+                        expect( mockUserActions.loginSuccess )
+                            .toHaveBeenCalledWith(user, date);
+                        expect( mockUserActions.loginFailed )
+                            .not.toHaveBeenCalled();
+                        done();
```

Comment:
It's generally best to stick to one expectation per test, it helps keep our tests focused and concise, and isolated

In this case, we might think about this test a little differently
Rather than trying to assert that a successful API call means that `mockLoginService.login` has been called, and also that `mockUserActions.loginSuccess` was called and also `mockUserActions.loginFailed` was _not_ called

(This more tightly couples our tests to specific implementation details) - we might instead think

One test might be, when an action of type `UserActions.LOGIN_USER_REQUEST` is emitted, I want to assert that a login request is attempted. i.e. (`expect( mockLoginService.login ).toHaveBeenCalledWith(user, pass);`)

Another test might be, when a login request succeeds, I want to assert the appropriate success action is emitted. i.e. `expect( mockUserActions.loginSuccess ).toHaveBeenCalledWith(user, date);`
^ This specific expectation however only gets us so far... just because `mockUserActions.loginSuccess` is called, does not guarantee that our epic is working as expected. We could be invoking this action creator, but forgot to map it out during a refactor - this test would still pass.
Rather... it's better that we test the resulting action emission to determine that this success behaviour of the epic is correct.
The more meaningful outcome - ultimately the fact that we call `mockUserActions.loginSuccess` is an implementation detail

Another way to test this might be...

```typescript
userLoginEpic.login(action$).subscribe(action => {
  const expectedAction = mockUserActions.loginSuccess(user, date);
  expect(action).toEqual(expectedAction);
  done();
})
```